kv/kvserver: TestLossQuorumCauseLeaderlessWatcherToSignalUnavailable failed

kv/kvserver.TestLossQuorumCauseLeaderlessWatcherToSignalUnavailable [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsIbmCloudLinuxS390x_UnitTestsS390x/20986142?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsIbmCloudLinuxS390x_UnitTestsS390x/20986142?buildTab=artifacts#/) on release-25.4 @ [911c2e758fdc320bbf8eed67b2541c11bb8f9c0f](https://github.com/cockroachdb/cockroach/commits/911c2e758fdc320bbf8eed67b2541c11bb8f9c0f):


```
=== RUN   TestLossQuorumCauseLeaderlessWatcherToSignalUnavailable
    test_log_scope.go:171: test logs captured to: /artifacts/tmp/_tmp/bb83881c24a4b108656da9268f863369/logTestLossQuorumCauseLeaderlessWatcherToSignalUnavailable3081390625
    test_log_scope.go:82: use -show-logs to present logs inline
    test_server_shim.go:561: DRPC is enabled (override by TestingGlobalDRPCOption)
    client_replica_test.go:3061: 
        	Error Trace:	pkg/kv/kvserver/client_replica_test.go:3061
        	Error:      	Received unexpected error:
        	            	context deadline exceeded
        	Test:       	TestLossQuorumCauseLeaderlessWatcherToSignalUnavailable
    panic.go:635: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/bb83881c24a4b108656da9268f863369/logTestLossQuorumCauseLeaderlessWatcherToSignalUnavailable3081390625
--- FAIL: TestLossQuorumCauseLeaderlessWatcherToSignalUnavailable (19.56s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/kv-triage
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestLossQuorumCauseLeaderlessWatcherToSignalUnavailable.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-58790

--- a/pkg/kv/kvserver/client_replica_test.go
+++ b/pkg/kv/kvserver/client_replica_test.go
@@ -3044,21 +3044,19 @@ func TestLossQuorumCauseLeaderlessWatcherToSignalUnavailable(t *testing.T) {
 		return nil
 	})
 
-	sendPutRequestWithTimeout := func(repl *kvserver.Replica, timeout time.Duration) (*kvpb.Error, error) {
-		ctx, cancel := context.WithTimeout(ctx, timeout)
-		defer cancel()
+	sendPutRequest := func(repl *kvserver.Replica) *kvpb.Error {
 		ba := &kvpb.BatchRequest{}
 		ba.RangeID = desc.RangeID
 		ba.Timestamp = repl.Clock().Now()
 		ba.Add(putArgs(key, []byte("foo")))
 		_, pErr := repl.Send(ctx, ba)
-		return pErr, ctx.Err()
+		return pErr
 	}
 
 	// Requests should immediately return an error indicating that the range is
-	// unavailable.
-	pErr, ctxErr := sendPutRequestWithTimeout(repl, 2*time.Second)
-	require.NoError(t, ctxErr)
+	// unavailable. No timeout is used here because the leaderless watcher should
+	// cause the request to fail fast. See #161316.
+	pErr = sendPutRequest(repl)
 	require.Regexp(t, "replica has been leaderless for 10s", pErr)
 	require.True(t, errors.HasType(pErr.GoError(), (*kvpb.ReplicaUnavailableError)(nil)),
 		"expected ReplicaUnavailableError, got %v", err)
@@ -3082,15 +3080,9 @@ func TestLossQuorumCauseLeaderlessWatcherToSignalUnavailable(t *testing.T) {
 	testutils.SucceedsSoon(t, func() error {
 		for i := range numServers {
 			repl := tc.GetFirstStoreFromServer(t, i).LookupReplica(roachpb.RKey(key))
-			pErr, ctxErr = sendPutRequestWithTimeout(repl, 2*time.Second)
-			if ctxErr == nil && pErr == nil {
+			if pErr = sendPutRequest(repl); pErr == nil {
 				return nil
 			}
-		}
-		// If we reach this point, we know that the request failed, return the
-		// error.
-		if ctxErr != nil {
-			return ctxErr
 		}
 		return pErr.GoError()
 	})

